### ISSUE TITLE
Default shipping method by channel selectable from the admin

### ISSUE THREAD

**Issue by gorkalaucirica** [Sylius/Sylius#7178]
It would be nice to have a default shipping method selector for each channel in channel edit page.

Right now default shipping method is selected in [DefaultShippingMethodResolver](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Component/Core/Resolver/DefaultShippingMethodResolver.php) selecting the first shipping method returned by the repo.

I leave this issue as a reminder for myself, I want to finish other opened PR before I start with this one. If someone wants to develop it let me know to avoid working twice on the same issue

**Comment by pjedrzejewski** [comment]
Default shipping method is not that simple, because it may depend on the products that are in cart. I was thinking about introducing default shipping zone per channel, just like we have for taxes. So the default shipping method would be resolved based on the default zone, if customer is not logged in or has no default address. WDYT?

**Comment by gorkalaucirica** [comment]
Yes, seems a better approach, but maybe we should also take into account other strategies.

Why?

**Long explanation:**
I brought this issue because the customers of one of our sites were confused by the shipping costs shown in the cart. Sylius by default shows one shipping price (as mentioned above) that can be related or not to the method that the customer will choose later. Users see a total price that may change after shipping method selection during checkout and can increase a lot specially for international shipments. Customers get confused.

**TL;DR:** Customers get confused by the "random" shipping costs in cart

**What we did:** Create a fake shipping method and create a custom default shipping resolver, identify it and show a message saying that cost will be calculated based on the address. 

Maybe, having different shipping cost strategies when no address selected, and allowing to select the by channel would be interesting. Some strategies:

- Showing default shipping method based on the default zone as @pjedrzejewski pointed out
- No shipping costs until address selected
- Any other strategy (make it extendable)

This as may be a huge change in the codebase and can be tricked easily (as we did, not very clean though), but I thing is a nice to have feature in the middle term.

Hope this gives an insight of some use cases that can happen (I guess they are quite common).

Maybe we can work in that direction

**Comment by marmotz** [comment]
In fact, it's not a feature request, but a bug:

Return the first enabled shipping method does not work if the first shipping method was not in current channel:
error in ajax/cart/add "{code: 500, message: "Notice: Undefined index: xxx"} caused by an undefined index error in Sylius\Component\Core\Shipping\Calculator\FlatRateCalculator:calculate() line 33

As a quick fix, you should add channel in DefaultShippingMethodResolver::getDefaultShippingMethod() criterias.

**Comment by 11mb** [comment]
The Stale bot closed this issue, shouldn't this be reopened, and tell the stale bot to leave this open? I think this is indeed strange behaviour in Sylius: just picking the first method; What @pjedrzejewski is proposing seems like a better idea. When someone has the more advanced usecase when it depends (also) on the products in the cart this is something they can implement themselves. I'm going to take the same approach @gorkalaucirica did.

**Comment by marcis101** [comment]
> Indeed, picking the default method is quite a weird behaviour, I'd love to have an option not to have any shipping/payment method until the customer explicitly chooses them.

But that means you can't initially show (default) shipping costs to your customers => it becomes a surprise => they might leave

**Comment by jbator** [comment]
Is there a simple way to disable default shipping method?

**Comment by bakome** [comment]
Maybe for start it will be nice this default check to follow order by position of shipping methods?

**Comment by vvasiloi** [comment]
For my current project, I added a default shipping country per channel and adjusted everything around that, which means new resolvers and other services that operate with a country instead of a zone.
However, initially it was a default shipping zone, but I simplified it because it was more appropriate for the project, so I support Pawel's suggestion:

> I was thinking about introducing default shipping zone per channel, just like we have for taxes. So the default shipping method would be resolved based on the default zone, if customer is not logged in or has no default address. WDYT?

**Comment by Bastienbretenet** [comment]
For those who want to choose the default shipping you can simply [customize ShippingMethodRepository](https://docs.sylius.com/en/latest/customization/repository.html) and rewrite the createEnabledForChannelQueryBuilder function : 

Add to createQueryBuilder :

`->addOrderBy('o.position', 'ASC')`

Then change the position of shipping method in the back office. The position 0 which will be the default shipping.

**Comment by Bastienbretenet** [comment]
As for how sylius works, it seems to me not adapt to a classic shop. User know that he choose their shipping method after entering their address.
But the estimate of the shipping costs distorts the cart. I understand the value of an estimate but I see two problematic scenarios:

- I have a free shipping method (pick and collect) and another with charges: 
The cart can show shipping costs which can deter the order while the customer can choose a free method.

- The cart may display a false estimate in the case where the customer orders from another zone which has a higher (or lower) price. But user will see this when he choose the shipping method..

In any case, I would rather not display anything rather than display a questionable estimate.

For those who want to change the behavior I think it is possible to move the createNewOrderShipment function (OrderShipmentProcessor.php) when the user enters their address (function saveAddresses ?)... this is just an idea, not testing.

Thank you all for your comments!

**Comment by TheMilek** [comment]
Hi!
I'm closing this issue, as now since Sylius 1.12 you can choose the default shipping method by position [REF](https://github.com/Sylius/Sylius/pull/14940) 😅 
Cheers! 🍻
